// hardhat/plugins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HARDHAT_NETWORK_NAME = exports.lazyFunction = exports.lazyObject = exports.NomicLabsHardhatPluginError = exports.HardhatPluginError = exports.ProviderWrapper = void 0;
var wrapper_1 = require("./internal/core/providers/wrapper");
Object.defineProperty(exports, "ProviderWrapper", { enumerable: true, get: function () { return wrapper_1.ProviderWrapper; } });
var errors_1 = require("./internal/core/errors");
Object.defineProperty(exports, "HardhatPluginError", { enumerable: true, get: function () { return errors_1.HardhatPluginError; } });
Object.defineProperty(exports, "NomicLabsHardhatPluginError", { enumerable: true, get: function () { return errors_1.NomicLabsHardhatPluginError; } });
var lazy_1 = require("./internal/util/lazy");
Object.defineProperty(exports, "lazyObject", { enumerable: true, get: function () { return lazy_1.lazyObject; } });
Object.defineProperty(exports, "lazyFunction", { enumerable: true, get: function () { return lazy_1.lazyFunction; } });
var constants_1 = require("./internal/constants");
Object.defineProperty(exports, "HARDHAT_NETWORK_NAME", { enumerable: true, get: function () { return constants_1.HARDHAT_NETWORK_NAME; } });
