// @nomicfoundation/hardhat-network-helpers/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.reset = exports.takeSnapshot = exports.stopImpersonatingAccount = exports.setNextBlockBaseFeePerGas = exports.setStorageAt = exports.setPrevRandao = exports.setNonce = exports.setCoinbase = exports.setCode = exports.setBlockGasLimit = exports.setBalance = exports.impersonateAccount = exports.getStorageAt = exports.dropTransaction = exports.mineUpTo = exports.mine = exports.time = void 0;
__exportStar(require("./loadFixture"), exports);
exports.time = __importStar(require("./helpers/time"));
var mine_1 = require("./helpers/mine");
Object.defineProperty(exports, "mine", { enumerable: true, get: function () { return mine_1.mine; } });
var mineUpTo_1 = require("./helpers/mineUpTo");
Object.defineProperty(exports, "mineUpTo", { enumerable: true, get: function () { return mineUpTo_1.mineUpTo; } });
var dropTransaction_1 = require("./helpers/dropTransaction");
Object.defineProperty(exports, "dropTransaction", { enumerable: true, get: function () { return dropTransaction_1.dropTransaction; } });
var getStorageAt_1 = require("./helpers/getStorageAt");
Object.defineProperty(exports, "getStorageAt", { enumerable: true, get: function () { return getStorageAt_1.getStorageAt; } });
var impersonateAccount_1 = require("./helpers/impersonateAccount");
Object.defineProperty(exports, "impersonateAccount", { enumerable: true, get: function () { return impersonateAccount_1.impersonateAccount; } });
var setBalance_1 = require("./helpers/setBalance");
Object.defineProperty(exports, "setBalance", { enumerable: true, get: function () { return setBalance_1.setBalance; } });
var setBlockGasLimit_1 = require("./helpers/setBlockGasLimit");
Object.defineProperty(exports, "setBlockGasLimit", { enumerable: true, get: function () { return setBlockGasLimit_1.setBlockGasLimit; } });
var setCode_1 = require("./helpers/setCode");
Object.defineProperty(exports, "setCode", { enumerable: true, get: function () { return setCode_1.setCode; } });
var setCoinbase_1 = require("./helpers/setCoinbase");
Object.defineProperty(exports, "setCoinbase", { enumerable: true, get: function () { return setCoinbase_1.setCoinbase; } });
var setNonce_1 = require("./helpers/setNonce");
Object.defineProperty(exports, "setNonce", { enumerable: true, get: function () { return setNonce_1.setNonce; } });
var setPrevRandao_1 = require("./helpers/setPrevRandao");
Object.defineProperty(exports, "setPrevRandao", { enumerable: true, get: function () { return setPrevRandao_1.setPrevRandao; } });
var setStorageAt_1 = require("./helpers/setStorageAt");
Object.defineProperty(exports, "setStorageAt", { enumerable: true, get: function () { return setStorageAt_1.setStorageAt; } });
var setNextBlockBaseFeePerGas_1 = require("./helpers/setNextBlockBaseFeePerGas");
Object.defineProperty(exports, "setNextBlockBaseFeePerGas", { enumerable: true, get: function () { return setNextBlockBaseFeePerGas_1.setNextBlockBaseFeePerGas; } });
var stopImpersonatingAccount_1 = require("./helpers/stopImpersonatingAccount");
Object.defineProperty(exports, "stopImpersonatingAccount", { enumerable: true, get: function () { return stopImpersonatingAccount_1.stopImpersonatingAccount; } });
var takeSnapshot_1 = require("./helpers/takeSnapshot");
Object.defineProperty(exports, "takeSnapshot", { enumerable: true, get: function () { return takeSnapshot_1.takeSnapshot; } });
var reset_1 = require("./helpers/reset");
Object.defineProperty(exports, "reset", { enumerable: true, get: function () { return reset_1.reset; } });
